[Inserter]: Add keyboard navigation in Patterns

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Resolves: https://github.com/WordPress/gutenberg/issues/28245

This PR adds keyboard navigation to the `Patterns List` in inserter. It uses the `Composite` component from `Reakit`, which is also used for block types list.


## Testing instructions
1. Open the inserter and navigate with keyboard through `Patterns`.
2. Everything else should be identical with the current implementation (styling).
<!-- Please describe what you have changed or added -->

ntsekouras committed:
[Inserter]: Add keyboard navigation in Patterns

ntsekouras committed:
add aria-label and update the readme

ntsekouras committed:
update e2e tests

diegohaz commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -43,18 +38,34 @@ function BlockPattern( { isDraggable, pattern, onClick } ) {
+						{ ...composite }
+						className="block-editor-block-patterns-list__item"
+						onClick={ () => onClick( pattern, blocks ) }
+						onKeyDown={ ( event ) => {
+							if (
+								ENTER === event.keyCode ||
+								SPACE === event.keyCode
+							) {
+								onClick( pattern, blocks );
+							}
+						} }
```
I don't think this `onKeyDown` is needed anymore as `CompositeItem` already handles this.
```suggestion
```

diegohaz reviewed (commented):
Thanks @ntsekouras! I made a suggestion, but the code looks good overall. I wonder if we need design feedback for the layout change?

youknowriad commented:
yeah, can't we retain the two columns layout for the search results?

ntsekouras committed:
address feedback

ntsekouras commented:
Thanks for reviewing @diegohaz and @youknowriad 👍  - I reverted the styles and addressed the feedback. I've also added the `orientation` as an attribute for having better key navigation in patterns from search results, as they are in grid.

mcsf commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -43,18 +37,26 @@ function BlockPattern( { isDraggable, pattern, onClick } ) {
-					) }
+					<CompositeItem
+						role="option"
+						as={ 'div' }
```
```suggestion
						as="div"
```

mcsf reviewed (approved):
Code looks good, keyboard interaction looks good!

youknowriad reviewed (approved):
LGTM 👍

ntsekouras committed:
review feedback

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>
